### PR TITLE
[RFC 0088] Nixpkgs Breaking Change Policy

### PR DESCRIPTION
**Pull request by kevincox**
This specifies the expected workflow as a base for workflows and future automation.

[Rendered](https://github.com/kevincox/rfcs-1/blob/patch-1/rfcs/0088-nixpkgs-breaking-change-policy.md)

### CODE REVIEW

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (SuperSandro2000) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:43]: How? Should I collect email addresses and write them an email? I think this is not practically possible.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (SuperSandro2000) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:65]: In certain cases the package could also be dropped. I feel like this should be mentioned here.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (SuperSandro2000) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:75]: I have the fear that this link will bit rot over time.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:43]: The idea was to tag them on the PR. We have contact info in https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix so this can (and is expected to be) automated.

Comment (Ma27) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:43]: I know this pops up quite often recently (or maybe I'm wrong and it's just selective perception), but this is something that can be solved by Hydra notifications or e.g. https://github.com/NixOS/hydra/pull/743.

Also I'd like to emphasize that there's a big difference between the stability of `master` and release branches, I mean there's a reason why `nixos-unstable` has `unstable` in its name and we have ZHF iterations after branch-off.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (FRidh) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:17]: The initial target branch that is chosen depends on several factors
- awareness of the staging branch and what its for
- knowing the amount of rebuilds needed and thereby whether it should target staging or master
- whether requires changes are on any of the possible target branches

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:65]: We discussed this and there was much disagreement of when to drop vs mark broken. I think that dropping because the maintainer doesn't respond in 7 days is a bit aggressive (they could just be on vacation) so I think the best route is just make it a simple "mark broken" and we can leave when to delete derivations to another policy.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:75]: Do you have a preferred link? I could just leave it as a description but I figured the link didn't harm.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (FRidh) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:40]: Whether it is sometimes necessary to break all dependent packages is quite a claim and needs further motivation.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (FRidh) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:42]: Why recommendation and not required?

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (Ma27) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:65]: If I understand Sandro correctly here, this isn't supposed to be the "default measure", but a valid thing to do if the package is e.g. fairly old.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (FRidh) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:43]: GitHub also limits the amount of mentions in case many people are affected.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (FRidh) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:50]: This assumes here that the change is sufficiently small it can go into master. With large changes that motivate this PR that is exactly not the case. The changes then go into staging, then staging-next and finally in master. That also gives an extra stabilization period on top of the proposed 7 days.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:43]: Part of the idea here is to allows us to move towards keeping the unstable branches green (at least from the building point of view). This allows security fixes and other changes to get pushed at any time. This policy won't get us all of the way there but it already in theory does as all broken packages should be marked as broken before getting merged to `master`.

Hydra notifications would be good, but they only apply to the merged branch. At that point it is too late, that broken package is already blocking changes which is bad for critical fixes and annoying to contributors (as their change gets delayed). Ideally this would be the same sort of tech, but applied somewhat pre-merge.

I don't know if there is an official stance on what `-unstable` means. Because the channel is always tested on hydra before being published. My understanding it is more that packages, including major versions will be updated frequently. I don't think it should mean "fails to build and release for days or weeks at a time".

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (FRidh) [inline rfcs/0000-nixpkgs-breaking-change-policy.md:70]: What is the scope of the change we are talking about? Staging is not for testing or dumping changes in the hope they get fixed; one is expected to follow their change and fix regressions especially when it has reached staging-next.

There are important NixOS tests and tests that are pretty unimportant.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (Ma27) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:43]: Well, the channel doesn't break on **any** package though. Most of the leaf-packages (and AFAIU that's what we're talking about here) can be totally broken (i.e. failing to compile or whatever) without preventing `nixos-unstable` from moving forward.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:65]: That makes sense to me, but I just wanted to keep it simple. In this policy the action for every package is simply to mark as broken. Then we can have some separate process for removing broken packages. I think trying to work a policy in here for when to delete is just unnecessary complication.

Note also that the intention is to create a tool to add the `broken = true` flags so we don't want to need to teach that tool to make this decision. It would be easiest to create a separate tool to raise PRs for old broken packages that will catch all packages, no matter why they were originally marked as broken.

**Review by FRidh — CHANGES_REQUESTED** [review]
Good initiative to take this further. Please do limit the line length to make it easier to comment.

The RFC does not go sufficiently into size/scope of PR's this RFC covers nor how it relates to the different development branches (master/staging/staging-next) we have. Motivations are missing for the specific steps in the procedure. Maybe they should be added there or in a separate section.

When talking about what is an important and unimportant change it would be very helpful to first have defined what is considered core and what not. I think it is difficult to define the policy well in this RFC without defining package tiers.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:43]: That's a good point. For what I am aiming for I don't really care about a package if the channel config doesn't care about it. On the other hand is there a downside to marking these broken packages as broken? If there is it would be easy enough to update the RFC to define what dependents we care about and just ignore them completely (maybe still notify them, but no need to mark as broken).

**Comment by kevincox** [comment]
> Good initiative to take this further. Please do limit the line length to make it easier to comment.

If that is desired I can hard-wrap it on the next push.

> The RFC does not go sufficiently into size/scope of PR's this RFC covers nor how it relates to the different development branches (master/staging/staging-next) we have. Motivations are missing for the specific steps in the procedure. Maybe they should be added there or in a separate section.

Right now I am focused on `master` and the `-unstable` channels which follow it. I am welcome to feedback on how this should apply to other branches but I am less familiar with them, (although IIUC they already require stability)
 
> When talking about what is an important and unimportant change it would be very helpful to first have defined what is considered core and what not. I think it is difficult to define the policy well in this RFC without defining package tiers.

This is being discussed here: https://github.com/NixOS/rfcs/pull/88#discussion_r593750923

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:40]: I should probably remove the "all" but I'm sure there are cases where libraries make updates that change the build system or remove something. The point here is that the initial review should consider if the amount of breakage is reasonable. I will reword to make that more explicit.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:50]: Good point. I will update the target branch to be variable as it should still follow the correct procedures with that regard.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:17]: This is largely what this policy is trying to address. We are trying to write down what should be done to make these choices better known and more clear.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:50]: For reference I believe this is the current policy for picking a target branch: https://github.com/NixOS/rfcs/blob/master/rfcs/0026-staging-workflow.md#detailed-design

Comment (FRidh) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:50]: That's the RFC. The implementation is in the Nixpkgs manual, https://nixos.org/manual/nixpkgs/stable/#submitting-changes-commit-policy

Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md:50]: Ah, thanks!

**Comment by kevincox** [comment]
Updated:

* Hard wrap lines.
* Remove many instances of `master` and refer to "target branch" according to existing policy for picking the target branch.
* Update number in file name to the PR number.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (SuperSandro2000) [reply rfcs/0000-nixpkgs-breaking-change-policy.md]: > We have contact info in

Those information are optional and not all people added a github name or an email address.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (SuperSandro2000) [reply rfcs/0000-nixpkgs-breaking-change-policy.md]: I have mirrored it to web.archive.org https://web.archive.org/web/20210325015950if_/https://paper.dropbox.com/doc/nixpkgs-Merge-Queue-Overview-MTSY8xKH6y1xDEwavyDNW

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md]: Well if they aren't contactable than their package will be broken. I don't think we can really do anything about this if we don't have any handles to the maintainer.

Comment (SuperSandro2000) [reply rfcs/0000-nixpkgs-breaking-change-policy.md]: We should add something similar to ``if you want to be contacted you need to add contact information because otherwise it is not possible``.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [inline rfcs/0088-nixpkgs-breaking-change-policy.md:74]:
```suggestion
packages (herein called sub-maintainers).
    1. The sub-maintainers will be contacted preferably by the `github` field,
      followed by the `email` field of [maintainer-list.nix
      ](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix).
      If neither of those fields are set than contacting the maintainer is not
      required and the package should be marked as broken after 7 days.
```

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md]: How does this sound? https://github.com/NixOS/rfcs/pull/88/files#r602721463

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:74]: Should we wait? Or remove this maintainer from the blocking list right away? What is the odds that they randomly notice that we are breaking their package?

**Review by sternenseemann — COMMENTED** [review]
Overall, I'd say what this RFC lacks most of all is a clear definition of a “breaking change”. From reading it, I get the impression that it is used as “a change which causes a build failure in a reverse dependency”. Clarification of this would be important in my eyes.

Very important would be to clarify that this policy doesn't apply to the introduction of (partial) evaluation errors by e. g. changing an internal interface since we can't work around such a problem using meta attributes and the channel would be held back until all breakage would be fixed.

Additionally, I think we can not just ignore breakage for users, as outlined in a comment.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (sternenseemann) [reply rfcs/0000-nixpkgs-breaking-change-policy.md]: I think removing old, unmaintained packages should be done independently of major updates of packages they depend on. Adding this as a valid option in the process adds extra need for discussion which is hindering if you are aiming to keep a defined schedule.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:78]: I think that for the purposes here we should consider the process of "how a patch is submitted" off-topic. If our current process is incapable of hitting the 7 day deadline then we should probably make improvements to that process (or can pick a longer deadline here). One other option is that there is a no-response deadline and an extension for in-progress fixes. For example 7 days to commit to a fix and 14 days to have the fix submitted. If all responses are merged in 7 days there is no need to wait longer.

But personally I don't feel the need to change this at this point. Also I explicitly said **at least** 7 days for the merge. The maintainer should use their own discretion.

> Additionally there should probably be a standardized way of keeping track of such PRs to help with this process.

How about suggesting that the PR description (and probably at least one of the commit messages) should link to the upstream PR? That way GitHub will link them together.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:87]: I am slightly hesitant to giving special privileges to changes that require dependent updates but maybe this is justified as packages with more dependents are likely more important.

I am also hesitant to put the full burden of merging these changes on an individual.

I guess this is blocking if this policy puts additional difficulties to getting the change merged. But as I see the additional burden is on the author/maintainer, not on the actual process of getting it merged.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:121]: FWIW I have a very early "prototype" here. It needs proper testing to see how effective this solution is. I can work on finishing it up and propose both in parallel if you think it is critical.

I considered it future work because I think this policy is still effective even if it isn't automated. But maybe that is too much burden to impose temporarily.

Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:121]: I think it would have to be building dependents and NixOS tests. Ideally it would be Hydra configured the same as the the channels that follow the eventual target branch. I think the main concern with Hydra is additional load.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:122]: That's great. It looks similar to what I had sketched out. I guess I can throw away my design and just make some tweaks to that script.

https://paper.dropbox.com/doc/nixpkgs-Merge-Queue-Easy-Mark-Broken--BHsIMZ2cLiBAvMfU66U0H9RIAg-A232tyO0zLDbUZ7zhAefX

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:16]: The fundamental reason I excluded this is that I don't think it is feasible to expect maintainers to know about all of the breakages that their change can cause in the dependent. At the end of the day I think the best we can do is say "If you liked it then you should have put a test on it".

**Comment by spacekookie** [comment]
The RFC is open for shepherd nominations

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (asymmetric) [inline rfcs/0088-nixpkgs-breaking-change-policy.md:97]: Newline missing
```suggestion
# Alternatives

## Shorter grace period
```

**Comment by lheckemann** [comment]
Nominations for the shepherd team are welcome! @kevincox maybe you have someone in mind who you think could help move this forward?

**Comment by kevincox** [comment]
@blaggacao would you be interested and able?

**Comment by kevincox** [comment]
@blaggacao would you be interested and able?

**Comment by blaggacao** [comment]
I accept nomination.

**Comment by blaggacao** [comment]
@nrdxp you have valuable experience (bors, hydra) to further this RFC, would you accept nomination for Shepherd?

**Comment by Mic92** [comment]
We still need more nomination for this RFC to proceed.

File: rfcs/0000-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0000-nixpkgs-breaking-change-policy.md]: What are your suggestions? Should we have a list of tests that can't be marked broken? In that case I think we would need to block the initial PR,

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (jonringer) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:125]: This could be a part of https://github.com/jonringer/basinix

**Comment by jonringer** [comment]
I'll nominate myself as a shepard

**Comment by kevincox** [comment]
Thanks Jonathan!

**Comment by spacekookie** [comment]
Anyone else interested in being a shepherd for this RFC?

**Comment by kevincox** [comment]
@nrdxp Can you send me your Matrix ID? I sent an invite to the email you have listed in maintainers-list.nix but I don't think it got through.

**Comment by nrdxp** [comment]
yeah np, it's @timdeh:matrix.org

**Comment by kevincox** [comment]
I believe that we have the requisite number of shepherds now. @blaggacao do you wish to be lead-shepherd?

**Comment by blaggacao** [comment]
I can do that.

**Comment by kevincox** [comment]
Hi everyone. I just pushed an update to the RFC that attempts to more explicitly answer the most common questions here. Please let me know if you still have any unaddressed questions.

I have also written a long-form article about where I hope this RFC allows us to go. While I hope that this RFC is self-describing and useful on its own it may help others to see the big picture spelt out as well. Please let me know if it would be helpful to bring any context from this post into this RFC: https://kevincox.ca/2021/08/21/nixpkgs-ci/

**Comment by blaggacao** [comment]
Shepherds Meeting, Monday, 23th of August 2021

- How to practically derive new failures? @jonringer 
  - Open question for implementation of tooling. This RFC tries to agree on a policy that can then be implemented. @kevincox 
- Maybe start by suggesting a sample to build? @blaggacao 
  - 3 hours of build time / 100 packages. Only very general guidelines when interpreting the policy. Effectiveness of this policy depends heavily on tooling support. @kevincox
- Special role of passtru test? -> codify specific testing contracts in a complementary RFC. @jonringer 
  - e.g. "Define passtru tests to secure detection of a "breaking" change according to this RFC"
- Notification: Noise vs Signal ratio? @jonringer
  - Tooling will have to account for this problem in appropriate ways. The policy in this RFC ca be interpreted as a "principle right to be notified" under certain scenarios. @kevincox 
  - Maybe fundamentally conceptualize notification as a publisher endpoint that users can subscribe to and provide some optional additional notification middleware on top of that. @blaggacao 
- Automate detection of notifiable downstream packages shouldn't be too hard. @nrdxp
  - Though currently, none seem readily available.
- Until tooling is in place, it might be impractical to hard-enforce this policy. However, that does not hinder adoption of this RFC. @kevincox 
- Do we observe different message classes, such as "build failure", "test failure" that we should react/notify differently to? @blaggacao 
  - Possibly, but for the sake of simplicity they should all triggering the same 7 day grace period, if deemed eligible. @kevincox 


In addition to the above discussion the existing feedback was summarily reviewed. Most of it was reported as addressed. I'd encourage the parties that have given feedback in the past to revise the document under the light of the above discussion and the recent changes and potentially renew their questions / observations to the author.

Otherwise, in my opinion, we might be slowly approaching FCP.

**Comment by jonringer** [comment]
I think most contributors who have tried to push for large changes, and have needed to wait for input from many individuals will agree that some convention about a grace period is needed. In general, this should only affect PRs against staging, as most PRs against master can generally be tested more thoroughly, and the effects of the changes usually carry much lower risk of breaking downstream packages (because there's significantly less of them).

I think we should move this into FCP.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (Mic92) [inline rfcs/0088-nixpkgs-breaking-change-policy.md:6]:
```suggestion
shepherd-leader: @blaggacao
```

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (Zimmi48) [inline rfcs/0088-nixpkgs-breaking-change-policy.md:36]:
```suggestion
order to be covered by this policy please put tests on your packages.)
```

**Comment by lheckemann** [comment]
@jonringer has the shepherd team agreed to move this to FCP with disposition to accept? If so, let us know and we'll update the label.

**Comment by blaggacao** [comment]
@lheckemann I confirm the shepherd team has agreed to move this to FCP with disposition to accept.

**Comment by jonringer** [comment]
sorry, got buried under other notifications, yes. We would like to move forward, with disposition to accept.

**Comment by Mic92** [comment]
We are now in the Final Comment Period (FCP). Unless any major objections are raised, this will be accepted and merged in 10 days from now, on November 13th!

Please post any final comments on this pull request.

**Review by rnhmjoj — APPROVED** [review]
I agree with pretty much everything proposed here. It's good to have some formal guidelines and documentation to point users to. Two considerations:

1. I think the grade period shouldn't be applied strictly. It should be a guideline or made flexible depending on the importance of the change and the sub-package: if a sub-package is not essential but important (some popular library, application, etc.) I'd rather delay a bit the update than to break it.  

2. I fear this could increase the burden on maintainers of mid to large packages (I mean those between being a leaf and core package). Those maintaining core packages are probably already well equipped and eat mass-rebuilds at breakfast, but not everyone else. It would be good if ofBorg, nixpkgs-review or some other new tool would automate doing some of the step in the proposed workflow before this RFC is enforced:
    - detecting broken packages
    - producing a list of contacts to ping the sub-maintainers
    - keep track of who replied
    - mark packages as broken

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (jonringer) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:99]: I'm okay with committers and active contributors to use my server: https://github.com/jonringer/server-configuration . If you're worried about compute. Since I'm not opening this up to the general public, it's not a solution, just a mitigation; but the offer is there.

Obviously, this is guideline is meant "within expectations" of a user's ability. Having some sort `nixpkgs-review` does make it a lot easier for reviewers to merge.

Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:99]: This is why we added the time limit and called it a guideline. The idea is even if you only own a potato we hope that you can spare 3h of your potato or find someone else to do so for big changes. If you think 3h is still to muck to ask I'd be happy to consider other thresholds, it was picked mostly arbitrarily to provide a rough target.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (kevincox) [reply rfcs/0088-nixpkgs-breaking-change-policy.md:117]: Eventually it should be all packages. My not-so-secret goal is that one day every single package will be considered "channel blocking" unless it is marked broken, basically meaning that there are no unintentional breakages. 

It is expected that tooling will manage this. This policy is not going to be enforced until we have made it reasonably easy to comply. This RFC is a first step to agree agree upon a policy first, then we can build up the tooling to implement the policy.

**Comment by 7c6f434c** [comment]
Sounds like nothing FCP-breaking has happened.

File: rfcs/0088-nixpkgs-breaking-change-policy.md
Comment (Synthetica9) [inline rfcs/0088-nixpkgs-breaking-change-policy.md:77]:
```suggestion
  maintainers are arguably the most critical to Nixpkgs and their load needs
```

**Pull request merged**
